🐛  Android build fails on CI

### What were you trying to do?

Tried to have the 2.5.0 build Android on CI (Bitrise). Have to mention this is the first time trying to build react-native-vision-camera on CI. IOS builds went fine.

### What happened instead?

Build fails with:
`ninja: error: '../../../../build/third-party-ndk/hermes/jni/armeabi-v7a/libhermes.so', needed by '../../../../build/intermediates/cmake/release/obj/armeabi-v7a/libVisionCamera.so', missing and no known rule to make it`

Doesn't like relative paths?

### Relevant log output

```shell
> Task :react-native-vision-camera:generateJsonModelRelease
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]armeabi-v7a : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]armeabi-v7a : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]arm64-v8a : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]arm64-v8a : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]x86 : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]x86 : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]x86_64 : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
/Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt : C/C++ release[REDACTED]x86_64 : CMake Warning at /Users/vagrant/git/node_modules/react-native-vision-camera/android/CMakeLists.txt:122 (message):
  VisionCamera linking: FOR_HERMES=true
> Task :react-native-vision-camera:prepareHermes UP-TO-DATE
> Task :react-native-vision-camera:prepareJSC
> Task :react-native-vision-camera:createNativeDepsDirectories UP-TO-DATE
> Task :react-native-vision-camera:downloadBoost
Download https://github.com/react-native-community/boost-for-react-native/releases/download/v1.63.0-0/boost_1_63_0.tar.gz
> Task :react-native-vision-camera:prepareBoost
> Task :react-native-vision-camera:downloadDoubleConversion
Download https://github.com/google/double-conversion/archive/v1.1.6.tar.gz
> Task :react-native-vision-camera:prepareDoubleConversion
> Task :react-native-vision-camera:downloadFolly
Download https://github.com/facebook/folly/archive/v2020.01.13.00.tar.gz
> Task :react-native-vision-camera:prepareFolly
> Task :react-native-vision-camera:downloadGlog
Download https://github.com/google/glog/archive/v0.3.5.tar.gz
> Task :react-native-vision-camera:prepareGlog
> Task :react-native-vision-camera:prepareThirdPartyNdkHeaders
> Task :react-native-vision-camera:externalNativeBuildRelease FAILED
Build VisionCamera_armeabi-v7a
ninja: Entering directory `/Users/vagrant/git/node_modules/react-native-vision-camera/android/.cxx/cmake/release/armeabi-v7a'
ninja: error: '../../../../build/third-party-ndk/hermes/jni/armeabi-v7a/libhermes.so', needed by '../../../../build/intermediates/cmake/release/obj/armeabi-v7a/libVisionCamera.so', missing and no known rule to make it
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':react-native-vision-camera:externalNativeBuildRelease'.
> Build command failed.
  Error while executing process /usr/local/share/android-sdk/cmake/3.10.2.4988404/bin/ninja with arguments {-C /Users/vagrant/git/node_modules/react-native-vision-camera/android/.cxx/cmake/release/armeabi-v7a VisionCamera}
  ninja: Entering directory `/Users/vagrant/git/node_modules/react-native-vision-camera/android/.cxx/cmake/release/armeabi-v7a'
  
  ninja: error: '../../../../build/third-party-ndk/hermes/jni/armeabi-v7a/libhermes.so', needed by '../../../../build/intermediates/cmake/release/obj/armeabi-v7a/libVisionCamera.so', missing and no known rule to make it
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7.1/userguide/command_line_interface.html#sec:command_line_warnings
BUILD FAILED in 11m 55s
651 actionable tasks: 651 executed
Build task failed, error: exit status 1
```


### Device

Default Bitrise build (Xcode 12.4.x, on macOS 10.15.7 (Catalina))

### VisionCamera Version

2.5.0

### Additional information

- [ ] I am using Expo
- [X] I have read the [Troubleshooting Guide](https://mrousavy.github.io/react-native-vision-camera/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

--- a/android/build.gradle
+++ b/android/build.gradle
@@ -241,33 +241,6 @@ dependencies {
   implementation "androidx.exifinterface:exifinterface:1.3.3"
 }
 
-task extractAARHeaders {
-  doLast {
-    configurations.extractHeaders.files.each {
-      def file = it.absoluteFile
-      copy {
-        from zipTree(file)
-        into "$buildDir/$file.name"
-        include "**/*.h"
-      }
-    }
-  }
-}
-
-task extractJNIFiles {
-  doLast {
-    configurations.extractJNI.files.each {
-      def file = it.absoluteFile
-
-      copy {
-        from zipTree(file)
-        into "$buildDir/$file.name"
-        include "jni/**/*"
-      }
-    }
-  }
-}
-
 // third-party-ndk deps headers
 // mostly a copy of https://github.com/software-mansion/react-native-reanimated/blob/master/android/build.gradle#L115
 
@@ -287,8 +260,10 @@ def GLOG_VERSION = thirdPartyVersions["GLOG_VERSION"]
 def glog_file = new File(downloadsDir, "glog-${GLOG_VERSION}.tar.gz")
 
 task createNativeDepsDirectories {
-  downloadsDir.mkdirs()
-  thirdPartyNdkDir.mkdirs()
+  doLast {
+    downloadsDir.mkdirs()
+    thirdPartyNdkDir.mkdirs()
+  }
 }
 
 task downloadBoost(dependsOn: createNativeDepsDirectories, type: Download) {
@@ -384,25 +359,46 @@ task prepareGlog(dependsOn: downloadGlog, type: Copy) {
   }
 }
 
-task prepareHermes() {
-  def hermesPackagePath = findNodeModulePath(projectDir, "hermes-engine")
-  if (!hermesPackagePath) {
-    throw new GradleScriptException("Could not find the hermes-engine npm package", null)
+task prepareThirdPartyNdkHeaders {
+  if (!boost_file.exists()) {
+    dependsOn(prepareBoost)
   }
-
-  def hermesAAR = file("$hermesPackagePath/android/hermes-debug.aar")
-  if (!hermesAAR.exists()) {
-    throw new GradleScriptException("The hermes-engine npm package is missing \"android/hermes-debug.aar\"", null)
+  if (!double_conversion_file.exists()) {
+    dependsOn(prepareDoubleConversion)
   }
-
-  def soFiles = zipTree(hermesAAR).matching({ it.include "**/*.so" })
-
-  copy {
-    from soFiles
-    from "$reactNative/ReactAndroid/src/main/jni/first-party/hermes/Android.mk"
-    into "$thirdPartyNdkDir/hermes"
+  if (!folly_file.exists()) {
+    dependsOn(prepareFolly)
+  }
+  if (!glog_file.exists()) {
+    dependsOn(prepareGlog)
   }
 }
+
+prepareThirdPartyNdkHeaders.mustRunAfter createNativeDepsDirectories
+
+task prepareHermes() {
+  doLast {
+    def hermesPackagePath = findNodeModulePath(projectDir, "hermes-engine")
+    if (!hermesPackagePath) {
+      throw new GradleScriptException("Could not find the hermes-engine npm package", null)
+    }
+
+    def hermesAAR = file("$hermesPackagePath/android/hermes-debug.aar")
+    if (!hermesAAR.exists()) {
+      throw new GradleScriptException("The hermes-engine npm package is missing \"android/hermes-debug.aar\"", null)
+    }
+
+    def soFiles = zipTree(hermesAAR).matching({ it.include "**/*.so" })
+
+    copy {
+      from soFiles
+      from "$reactNative/ReactAndroid/src/main/jni/first-party/hermes/Android.mk"
+      into "$thirdPartyNdkDir/hermes"
+    }
+  }
+}
+
+prepareHermes.mustRunAfter prepareThirdPartyNdkHeaders
 
 task prepareJSC {
   doLast {
@@ -434,29 +430,47 @@ task prepareJSC {
   }
 }
 
-task prepareThirdPartyNdkHeaders {
-  if (!boost_file.exists()) {
-    dependsOn(prepareBoost)
-  }
-  if (!double_conversion_file.exists()) {
-    dependsOn(prepareDoubleConversion)
-  }
-  if (!folly_file.exists()) {
-    dependsOn(prepareFolly)
-  }
-  if (!glog_file.exists()) {
-    dependsOn(prepareGlog)
+prepareJSC.mustRunAfter prepareHermes
+
+task extractAARHeaders {
+  doLast {
+    configurations.extractHeaders.files.each {
+      def file = it.absoluteFile
+      copy {
+        from zipTree(file)
+        into "$buildDir/$file.name"
+        include "**/*.h"
+      }
+    }
   }
 }
+
+extractAARHeaders.mustRunAfter prepareJSC
+
+task extractJNIFiles {
+  doLast {
+    configurations.extractJNI.files.each {
+      def file = it.absoluteFile
+
+      copy {
+        from zipTree(file)
+        into "$buildDir/$file.name"
+        include "jni/**/*"
+      }
+    }
+  }
+}
+
+extractJNIFiles.mustRunAfter extractAARHeaders
 
 // pre-native build pipeline
 
 tasks.whenTaskAdded { task ->
-  if (task.name.contains('externalNativeBuild')) {
-    task.dependsOn(prepareJSC)
-    task.dependsOn(prepareHermes)
+  if (task.name == 'externalNativeBuildDebug' || task.name == 'externalNativeBuildRelease') {
     task.dependsOn(extractAARHeaders)
     task.dependsOn(extractJNIFiles)
+    task.dependsOn(prepareJSC)
+    task.dependsOn(prepareHermes)
     task.dependsOn(prepareThirdPartyNdkHeaders)
   }
 }

--- a/example/android/gradle.properties
+++ b/example/android/gradle.properties
@@ -10,7 +10,7 @@
 # Specifies the JVM arguments used for the daemon process.
 # The setting is particularly useful for tweaking memory settings.
 # Default value: -Xmx10248m -XX:MaxPermSize=256m
-# org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx2048m -XX:MaxPermSize=512m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
 
 # When configured, Gradle will run in incubating parallel mode.
 # This option should only be used with decoupled projects. More details, visit

